BTP-Workshops/WS03/part 1/w3_p1.cpp

```diff
@@ -0,0 +1,97 @@
+
+    std::cout << "1. Book tester >>> ===========================================================" << std::endl;
+    std::cout << "total books: " << numBooks << std::endl;
+    std::cout << "------------------------------------------------------------------------------" << std::endl;
+    for (auto i = 0u; i < numBooks-1; i++) {
+        std::cout << "| " << i + 1 << ".";
+        books[i].print(std::cout) << "|" << std::endl;;
+    }
+    std::cout << "| " << numBooks << ".";
+    std::cout << books[numBooks-1] << "|" << std::endl;
+    std::cout << "============================================================== <<< Book tester" << std::endl << std::endl;
+
+    std::cout << "2. Collection tester for int and 10 >>> ======================================" << std::endl;
+    {        // Collection tester for int-type
+        sdds::Collection<int, 10> icol;
+        for(auto i : ints)
+            icol += i; 
+        std::cout << "(smallest,largest) items: (" << icol.getSmallestItem() << "," << icol.getLargestItem() << ")" << std::endl;
+        std::cout << "size/capacity: " << icol.size() << "/" << icol.capacity() << std::endl;
+        std::cout << "Contents: ";
+        icol.print(std::cout);
+    }
+    std::cout << "========================================= <<< Collection tester for int and 10" << std::endl;
+    std::cout << std::endl;
+
+    std::cout << "3. Collection tester for double and 10 >>>====================================" << std::endl;
+    {// Collection tester for double-type
+        sdds::Collection<double, 10> dcol;
+        for (auto d : doubles)
+            dcol += d;
+        std::cout << "(smallest,largest) items: (" << dcol.getSmallestItem() << "," << dcol.getLargestItem() << ")" << std::endl;
+        std::cout << "size/capacity: " << dcol.size() << "/" << dcol.capacity() << std::endl;
+        std::cout << "Contents: ";
+        dcol.print(std::cout);
+    }
+    std::cout << "====================================== <<< Collection tester for double and 10" << std::endl;
+    std::cout << std::endl;
+
+    std::cout << "4. Collection tester for Book and 10 >>> =====================================" << std::endl;
+    {// Collection tester for Book-type, capacity 10
+        std::cout << "[After creating collection in empty state]" << std::endl;
+        sdds::Collection<sdds::Book, 10> bcol;
+        std::cout << "Book with smallest pages-to-chapters ratio (initial-default state): " << std::endl;
+        bcol.getSmallestItem().print(std::cout) << " |" << std::endl;
+        std::cout << "Book with largest pages-to-chapters ratio (initial-deault-state): " << std::endl;
+        bcol.getLargestItem().print(std::cout) << " |" << std::endl;
+        std::cout << "size/capacity: " << bcol.size() << "/" << bcol.capacity() << std::endl;
+    }
+    std::cout << "======================================== <<< Collection tester for Book and 10" << std::endl;
+    std::cout << std::endl;
+
+    return cout + endl;
+}
```

Comment:
Conclusion for the Program output: ...
Begin Testing the Program!
--------------------------

Command Line Arguments:
---------------------------------
  1: ws
---------------------------------

1. Book tester >>> ===========================================================
total books: 6
------------------------------------------------------------------------------
| 1.         Software Architecture for Busy Developers,7,174 | (624.000000)   |
| 2.      Low-Code Application Development with Appian,16,462 | (628.000000)   |
| 3.       Natural Language Processing with TensorFlow,11,514 | (646.000000)   |
| 4.                                   Learning DevOps,17,560 | (632.000000)   |
| 5.                            SQL for Data Analytics,9,540 | (660.000000)   |
| 6.  Deciphering Object-Oriented Programming with C++,21,594 | (628.000000)   |
============================================================== <<< Book tester

2. Collection tester for int and 10 >>> ======================================
(smallest,largest) items: (0,35)
size/capacity: 7/10
Contents: 25
30
20
15
25
5
35
========================================= <<< Collection tester for int and 10

3. Collection tester for double and 10 >>>====================================
(smallest,largest) items: (0.000000,9.100000)
size/capacity: 6/10
Contents: 7.800000
9.100000
3.400000
1.200000
2.300000
6.700000
====================================== <<< Collection tester for double and 10

4. Collection tester for Book and 10 >>> =====================================
[After creating collection in empty state]
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string: construction from null is not valid
==83270== 
==83270== Process terminating with default action of signal 6 (SIGABRT)
==83270==    at 0x57E9387: raise (in /usr/lib64/libc-2.17.so)
==83270==    by 0x57EAA77: abort (in /usr/lib64/libc-2.17.so)
==83270==    by 0x4EDB83A: __gnu_cxx::__verbose_terminate_handler() [clone .cold] (vterminate.cc:95)
==83270==    by 0x4EEA9E5: __cxxabiv1::__terminate(void (*)()) (eh_terminate.cc:48)
==83270==    by 0x4EEAA50: std::terminate() (eh_terminate.cc:58)
==83270==    by 0x4EEAC93: __cxa_throw (eh_throw.cc:98)
==83270==    by 0x4EDDE5D: std::__throw_logic_error(char const*) (functexcept.cc:70)
==83270==    by 0x402274: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string<std::allocator<char> >(char const*, std::allocator<char> const&) (basic_string.h:636)
==83270==    by 0x4028AB: sdds::Book::Book() (Book.cpp:19)
==83270==    by 0x402531: sdds::Collection<sdds::Book, 10u>::Collection() (Collection.h:57)
==83270==    by 0x401CDC: main (w3_p1_prof.cpp:86)
==83270==
==83270== HEAP SUMMARY:
==83270==     in use at exit: 74,149 bytes in 8 blocks
==83270==   total heap usage: 14 allocs, 6 frees, 74,384 bytes allocated
==83270==
==83270== 23 bytes in 1 blocks are still reachable in loss record 1 of 8
==83270==    at 0x4C2B87D: operator new(unsigned long) (vg_replace_malloc.c:472)
==83270==    by 0x4F81D2B: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char*>(char*, char*, std::forward_iterator_tag) (basic_string.tcc:225)   
==83270==    by 0x40291A: sdds::Book::Book(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int) (Book.cpp:25)
==83270==    by 0x40173D: main (w3_p1_prof.cpp:44)
==83270==
==83270== 42 bytes in 1 blocks are still reachable in loss record 2 of 8
==83270==    at 0x4C2B87D: operator new(unsigned long) (vg_replace_malloc.c:472)
==83270==    by 0x4F81D2B: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char*>(char*, char*, std::forward_iterator_tag) (basic_string.tcc:225)   
==83270==    by 0x40291A: sdds::Book::Book(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int) (Book.cpp:25)
==83270==    by 0x401640: main (w3_p1_prof.cpp:44)
==83270==
==83270== 44 bytes in 1 blocks are still reachable in loss record 3 of 8
==83270==    at 0x4C2B87D: operator new(unsigned long) (vg_replace_malloc.c:472)
==83270==    by 0x4F81D2B: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char*>(char*, char*, std::forward_iterator_tag) (basic_string.tcc:225)   
==83270==    by 0x40291A: sdds::Book::Book(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int) (Book.cpp:25)
==83270==    by 0x4016BF: main (w3_p1_prof.cpp:44)
==83270==
==83270== 45 bytes in 1 blocks are still reachable in loss record 4 of 8
==83270==    at 0x4C2B87D: operator new(unsigned long) (vg_replace_malloc.c:472)
==83270==    by 0x4F81D2B: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char*>(char*, char*, std::forward_iterator_tag) (basic_string.tcc:225)   
==83270==    by 0x40291A: sdds::Book::Book(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int) (Book.cpp:25)
==83270==    by 0x401680: main (w3_p1_prof.cpp:44)
==83270==
==83270== 49 bytes in 1 blocks are still reachable in loss record 5 of 8
==83270==    at 0x4C2B87D: operator new(unsigned long) (vg_replace_malloc.c:472)
==83270==    by 0x4F81D2B: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char*>(char*, char*, std::forward_iterator_tag) (basic_string.tcc:225)   
==83270==    by 0x40291A: sdds::Book::Book(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int) (Book.cpp:25)
==83270==    by 0x40177C: main (w3_p1_prof.cpp:44)
==83270==
==83270== 74 bytes in 1 blocks are still reachable in loss record 6 of 8
==83270==    at 0x4C2B87D: operator new(unsigned long) (vg_replace_malloc.c:472)
==83270==    by 0x4F2D0EC: allocate (new_allocator.h:147)
==83270==    by 0x4F2D0EC: std::string::_Rep::_S_create(unsigned long, unsigned long, std::allocator<char> const&) (cow_string.h:3602)
==83270==    by 0x4F10156: _S_construct<char const*> (cow_string.h:3161)
==83270==    by 0x4F10156: _S_construct<char const*> (cow_string.h:3146)
==83270==    by 0x4F10156: _S_construct_aux<char const*> (cow_string.h:2139)
==83270==    by 0x4F10156: _S_construct<char const*> (cow_string.h:2160)
==83270==    by 0x4F10156: basic_string (cow_string.h:605)
==83270==    by 0x4F10156: std::logic_error::logic_error(char const*) (cow-stdexcept.cc:87)
==83270==    by 0x4EDDE47: std::__throw_logic_error(char const*) (functexcept.cc:70)
==83270==    by 0x402274: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string<std::allocator<char> >(char const*, std::allocator<char> const&) (basic_string.h:636)
==83270==    by 0x4028AB: sdds::Book::Book() (Book.cpp:19)
==83270==    by 0x402531: sdds::Collection<sdds::Book, 10u>::Collection() (Collection.h:57)
==83270==    by 0x401CDC: main (w3_p1_prof.cpp:86)
==83270==
==83270== 144 bytes in 1 blocks are possibly lost in loss record 7 of 8
==83270==    at 0x4C2B15C: malloc (vg_replace_malloc.c:431)
==83270==    by 0x4EE997F: __cxa_allocate_exception (eh_alloc.cc:398)
==83270==    by 0x4EDDE31: std::__throw_logic_error(char const*) (functexcept.cc:70)
==83270==    by 0x402274: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string<std::allocator<char> >(char const*, std::allocator<char> const&) (basic_string.h:636)
==83270==    by 0x4028AB: sdds::Book::Book() (Book.cpp:19)
==83270==    by 0x402531: sdds::Collection<sdds::Book, 10u>::Collection() (Collection.h:57)
==83270==    by 0x401CDC: main (w3_p1_prof.cpp:86)
==83270==
==83270== 73,728 bytes in 1 blocks are still reachable in loss record 8 of 8
==83270==    at 0x4C2B15C: malloc (vg_replace_malloc.c:431)
==83270==    by 0x4EE72CD: pool (eh_alloc.cc:235)
==83270==    by 0x4EE72CD: __static_initialization_and_destruction_0 (eh_alloc.cc:373)
==83270==    by 0x4EE72CD: _GLOBAL__sub_I_eh_alloc.cc (eh_alloc.cc:456)
==83270==    by 0x400F9C2: _dl_init (in /usr/lib64/ld-2.17.so)
==83270==    by 0x4001179: ??? (in /usr/lib64/ld-2.17.so)
==83270==
==83270== LEAK SUMMARY:
==83270==    definitely lost: 0 bytes in 0 blocks
==83270==    indirectly lost: 0 bytes in 0 blocks
==83270==      possibly lost: 144 bytes in 1 blocks
==83270==    still reachable: 74,005 bytes in 7 blocks
==83270==                       of which reachable via heuristic:
==83270==                         stdstring          : 74 bytes in 1 blocks
==83270==         suppressed: 0 bytes in 0 blocks
==83270==
==83270== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
bash: line 1: 83270 Aborted                 /usr/local/valgrind-3.21.0/bin/valgrind -s --leak-check=full --show-leak-kinds=all --track-origins=yes ws
Script done, file is output_p1_stud.txt

Checking output:
In line number 18 of your output:
The output should be:
| 1.         Software Architecture for Busy Developers,7,174 | (24.857143)    |
                                                                ^
But your output is:
| 1.         Software Architecture for Busy Developers,7,174 | (624.000000)   |
                                                                ^

Unmatched character details:
The character in column 65 is supposed to be:
  [2] ASCII code(50)
but you printed
  [6] ASCII code(54)

Outputs don't match. Submission aborted!